### PR TITLE
feat: upgrade to react 18 - fix build #4

### PR DESCRIPTION
**Pull request by neatbyte-vnobis**
## Changes
Fourth batch of fixes for project build after updating to react 18.
https://www.notion.so/webiny/Upgrade-to-React-18-ceacd9bbfcf74b39a7ceb70943c3afa7

Packages: app-page-builder.

## How Has This Been Tested?
Manually.

## Documentation
None.

### CODE REVIEW

File: packages/app-page-builder/src/admin/plugins/installation.tsx [hunk 1]
```diff
@@ -50,7 +50,7 @@ const PBInstaller = ({ onInstalled }: PbInstallerProps) => {
     const [loading, setLoading] = useState(false);
     const [error, setError] = useState(null);
 
-    const onSubmit = useCallback(async form => {
+    const onSubmit = useCallback(async (form: any) => {
```
Comment (brunozoric) [inline packages/app-page-builder/src/admin/plugins/installation.tsx]: https://github.com/webiny/webiny-js/blob/05ea175641f259bdfae4c1a528e4cb5f415b3bd3/packages/form/src/types.ts#L80

basically, `form` is actually `data` and of `Record<string, any>` type.

File: packages/app-page-builder/src/admin/plugins/permissionRenderer/PageBuilderPermissions/PageBuilderPermissions.tsx [hunk 2]
```diff
@@ -49,7 +49,7 @@ export const PageBuilderPermissions = ({ value, onChange }: PageBuilderPermissio
     }, []);
 
     const onFormChange = useCallback(
-        formData => {
+        (formData: any) => {
```
Comment (brunozoric) [inline packages/app-page-builder/src/admin/plugins/permissionRenderer/PageBuilderPermissions/PageBuilderPermissions.tsx]: Same as my previous comment. Please dont use `any`. Yes, there are some `any` in the code, but we are trying to eliminate it.

File: packages/app-page-builder/src/admin/views/BlockCategories/IconPicker.tsx [hunk 3]
```diff
@@ -110,9 +110,8 @@ const IconPicker = ({ value, onChange, label, description, validation }: IconPic
     }, [mustRenderGrid]);
 
     const onFilterChange = useCallback(
-        (value, cb) => {
```
Comment (brunozoric) [inline packages/app-page-builder/src/admin/views/BlockCategories/IconPicker.tsx:113]: Why did you remove the `cb`?

File: packages/app-page-builder/src/admin/views/Categories/CategoriesForm.tsx [hunk 4]
```diff
@@ -118,7 +118,7 @@ const CategoriesForm = ({ canCreate }: CategoriesFormProps) => {
     const loading = [getQuery, createMutation, updateMutation].find(item => item.loading);
 
     const onSubmit = useCallback(
-        async formData => {
+        async (formData: any) => {
```
Comment (brunozoric) [inline packages/app-page-builder/src/admin/views/Categories/CategoriesForm.tsx]: Please put some type, even the `Record<string, any>` is better than just `any`.

File: packages/app-page-builder/src/admin/views/Menus/MenusDataList.tsx [hunk 5]
```diff
@@ -71,7 +71,7 @@ const PageBuilderMenusDataList = ({ canCreate }: PageBuilderMenusDataListProps)
     const { showConfirmation } = useConfirmationDialog();
 
     const filterMenus = useCallback(
-        ({ title, slug, description }) => {
+        ({ title, slug, description }: any) => {
```
Comment (brunozoric) [inline packages/app-page-builder/src/admin/views/Menus/MenusDataList.tsx]: Please put correct type.

File: packages/app-page-builder/src/admin/views/Menus/MenusDataList.tsx [hunk 6]
```diff
@@ -82,7 +82,7 @@ const PageBuilderMenusDataList = ({ canCreate }: PageBuilderMenusDataListProps)
     );
 
     const sortMenus = useCallback(
-        menus => {
+        (menus: any) => {
```
Comment (brunozoric) [inline packages/app-page-builder/src/admin/views/Menus/MenusDataList.tsx]: Please put correct type.

File: packages/app-page-builder/src/admin/views/Menus/MenusForm.tsx [hunk 7]
```diff
@@ -92,7 +92,7 @@ const MenusForm = ({ canCreate }: MenusFormProps) => {
     const loading = [getQuery, createMutation, updateMutation].find(item => item.loading);
 
     const onSubmit = useCallback(
-        async formData => {
+        async (formData: any) => {
```
Comment (brunozoric) [inline packages/app-page-builder/src/admin/views/Menus/MenusForm.tsx]: Please put correct type.

File: packages/app-page-builder/src/admin/views/PageTemplates/CreatePageTemplateDialog.tsx [hunk 8]
```diff
@@ -39,7 +39,7 @@ type CreatePageTemplateDialogProps = {
 const CreatePageTemplateDialog = ({ onClose, onSubmit }: CreatePageTemplateDialogProps) => {
     const [loading, setLoading] = useState(false);
     const submitForm = useCallback(
-        async data => {
+        async (data: any) => {
```
Comment (brunozoric) [inline packages/app-page-builder/src/admin/views/PageTemplates/CreatePageTemplateDialog.tsx]: Please put correct type.

File: packages/app-page-builder/src/editor/Editor.tsx [hunk 9]
```diff
@@ -52,6 +52,8 @@ export const Editor = ({ stateInitializerFactory }: EditorProps) => {
     );
 
     return (
+        // 'children' prop missing from DndProvider types before react-dnd v14.0.5.
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/Editor.tsx:55]: Can you update the react-dnd package?

File: packages/app-page-builder/src/editor/components/IconPicker.tsx [hunk 10]
```diff
@@ -186,9 +186,8 @@ const IconPicker = ({
     const gridWidth = useInSidebar ? 300 : 640;
 
     const onFilterChange = useCallback(
-        (value, cb) => {
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/components/IconPicker.tsx:189]: Why did you remove `cb`?

File: packages/app-page-builder/src/editor/helpers.ts [hunk 11]
```diff
@@ -194,7 +194,7 @@ export const removeElementId = (el: PbElement): PbElement => {
     return el;
 };
 
-export const createBlockElements = (name: string): PbEditorElement => {
+export const createBlockElements = (name?: string): PbEditorElement => {
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/helpers.ts:197]: Name should not be optional.

File: packages/app-page-builder/src/editor/plugins/defaultBar/components/ExportBlockButton/ExportBlockLoadingDialogContent.tsx [hunk 12]
```diff
@@ -31,11 +33,11 @@ interface ExportBlockLoadingDialogContent {
 
 const ExportBlockLoadingDialogContent = ({ taskId }: ExportBlockLoadingDialogContent) => {
     const [completed, setCompleted] = useState<boolean>(false);
-    const [error, setError] = useState<Error | null>(null);
+    const [error, setError] = useState<Record<string, string> | null>(null);
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/plugins/defaultBar/components/ExportBlockButton/ExportBlockLoadingDialogContent.tsx:36]: I would say that error is of type `{message: string, code: string, data: Record<string, any>}`. I saw that you defined it in a file which contained GraphQL response types. Maybe have some general `PbApiError` interface and use that everywhere?

File: packages/app-page-builder/src/editor/plugins/defaultBar/components/ExportPageButton/ExportPageLoadingDialogContent.tsx [hunk 13]
```diff
@@ -31,11 +33,11 @@ interface ExportPageLoadingDialogContent {
 
 const ExportPageLoadingDialogContent = ({ taskId }: ExportPageLoadingDialogContent) => {
     const [completed, setCompleted] = useState<boolean>(false);
-    const [error, setError] = useState<Error | null>(null);
+    const [error, setError] = useState<Record<string, string> | null>(null);
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/plugins/defaultBar/components/ExportPageButton/ExportPageLoadingDialogContent.tsx:36]: Same as in previous file.

File: packages/app-page-builder/src/editor/plugins/defaultBar/components/ExportTemplateButton/ExportTemplateLoadingDialogContent.tsx [hunk 14]
```diff
@@ -31,40 +33,46 @@ interface ExportTemplateLoadingDialogContent {
 
 const ExportTemplateLoadingDialogContent = ({ taskId }: ExportTemplateLoadingDialogContent) => {
     const [completed, setCompleted] = useState<boolean>(false);
-    const [error, setError] = useState<Error | null>(null);
+    const [error, setError] = useState<Record<string, string> | null>(null);
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/plugins/defaultBar/components/ExportTemplateButton/ExportTemplateLoadingDialogContent.tsx:36]: Error type.

File: packages/app-page-builder/src/editor/plugins/defaultBar/components/ExportTemplateButton/ExportTemplateLoadingDialogContent.tsx [hunk 15]
```diff
@@ -31,40 +33,46 @@ interface ExportTemplateLoadingDialogContent {
...
 
-        // Handler failed task
-        if (data && data.status === "failed") {
-            setCompleted(true);
-            showSnackbar("Error: Failed to export templates!");
-            setError(data.error);
-        }
+    const pollExportTemplateTaskStatus = useCallback(
+        (response: GetTemplateImportExportTaskResponse) => {
+            const { error, data } = response?.pageBuilder.getImportExportTask || {};
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/plugins/defaultBar/components/ExportTemplateButton/ExportTemplateLoadingDialogContent.tsx]: How is it possible that `response` is undefined?

File: packages/app-page-builder/src/editor/plugins/defaultBar/components/ImportButton/block/useImportBlockLoadingDialog.tsx [hunk 16]
```diff
@@ -37,9 +37,9 @@ interface ImportBlockLoadingDialogContentProps {
 const ImportBlockLoadingDialogContent = ({ taskId }: ImportBlockLoadingDialogContentProps) => {
     const { showSnackbar } = useSnackbar();
     const [completed, setCompleted] = useState<boolean>(false);
-    const [error, setError] = useState<Error | null>(null);
+    const [error, setError] = useState<Record<string, string> | null>(null);
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/plugins/defaultBar/components/ImportButton/block/useImportBlockLoadingDialog.tsx:40]: Error type.

File: packages/app-page-builder/src/editor/plugins/defaultBar/components/ImportButton/page/useImportPageLoadingDialog.tsx [hunk 17]
```diff
@@ -37,9 +38,9 @@ interface ImportPageLoadingDialogContentProps {
 const ImportPageLoadingDialogContent = ({ taskId }: ImportPageLoadingDialogContentProps) => {
     const { showSnackbar } = useSnackbar();
     const [completed, setCompleted] = useState<boolean>(false);
-    const [error, setError] = useState<Error | null>(null);
+    const [error, setError] = useState<Record<string, string> | null>(null);
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/plugins/defaultBar/components/ImportButton/page/useImportPageLoadingDialog.tsx:41]: Error type.

File: packages/app-page-builder/src/editor/plugins/defaultBar/components/ImportButton/template/useImportTemplateLoadingDialog.tsx [hunk 18]
```diff
@@ -39,17 +39,20 @@ const ImportTemplateLoadingDialogContent = ({
...
-    const { data } = useQuery(GET_TEMPLATE_IMPORT_EXPORT_TASK, {
-        variables: {
-            id: taskId
-        },
-        skip: taskId === null,
-        fetchPolicy: "network-only",
-        pollInterval: completed ? 0 : INTERVAL,
-        notifyOnNetworkStatusChange: true
-    });
+    const [error, setError] = useState<Record<string, string> | null>(null);
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/plugins/defaultBar/components/ImportButton/template/useImportTemplateLoadingDialog.tsx:42]: Error type.

File: packages/app-page-builder/src/editor/plugins/elementSettings/components/SelectField.tsx [hunk 19]
```diff
@@ -61,7 +61,6 @@ const SelectField = ({
     return (
         <React.Fragment>
             <select
-                placeholder={placeholder}
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/plugins/elementSettings/components/SelectField.tsx:64]: Why did you remove `placeholder`?

File: packages/app-page-builder/src/editor/plugins/elements/button/ButtonSettings.tsx [hunk 20]
```diff
@@ -75,8 +75,11 @@ const ButtonSettings = ({
         postModifyElement: updateButtonElementIcon
     });
 
-    const updateType = useCallback(value => getUpdateValue("type")(value), [getUpdateValue]);
-    const updateIcon = useCallback(value => getUpdateValue("icon.id")(value?.id), [getUpdateValue]);
+    const updateType = useCallback((value: any) => getUpdateValue("type")(value), [getUpdateValue]);
+    const updateIcon = useCallback(
+        (value: PbIcon) => getUpdateValue("icon.id")(value?.id),
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/plugins/elements/button/ButtonSettings.tsx]: How is value `undefined`? Your type says its always defined. No need to do `value?.id`.

File: packages/app-page-builder/src/editor/plugins/elements/button/ButtonSettings.tsx [hunk 21]
```diff
@@ -75,8 +75,11 @@ const ButtonSettings = ({
         postModifyElement: updateButtonElementIcon
     });
 
-    const updateType = useCallback(value => getUpdateValue("type")(value), [getUpdateValue]);
-    const updateIcon = useCallback(value => getUpdateValue("icon.id")(value?.id), [getUpdateValue]);
+    const updateType = useCallback((value: any) => getUpdateValue("type")(value), [getUpdateValue]);
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/plugins/elements/button/ButtonSettings.tsx]: I would say that `value` is a `string`.

File: packages/app-page-builder/src/editor/plugins/toolbar/addElement/AddElement.tsx [hunk 22]
```diff
@@ -76,6 +76,15 @@ const ElementsList = ({ groupPlugin, elements, renderDraggable, refresh }: Eleme
     );
 };
 
+interface RenderOverlayCb {
+    (
+        element: any,
```
Comment (brunozoric) [inline packages/app-page-builder/src/editor/plugins/toolbar/addElement/AddElement.tsx]: ... `any`?

File: packages/app-page-builder/src/pageEditor/helpers.ts [hunk 23]
```diff
@@ -9,7 +9,7 @@ import {
     PbEditorPageElementVariableRendererPlugin
 } from "~/types";
 
-export const createBlockReference = (name: string): PbEditorElement => {
+export const createBlockReference = (name?: string): PbEditorElement => {
```
Comment (brunozoric) [inline packages/app-page-builder/src/pageEditor/helpers.ts:12]: `name` can really be `undefined`?

**Review by brunozoric — CHANGES_REQUESTED** [review]

File: packages/app-page-builder/src/admin/views/BlockCategories/IconPicker.tsx (hunk 3 above)
Comment (neatbyte-ivelychko) [reply packages/app-page-builder/src/admin/views/BlockCategories/IconPicker.tsx:113]: Previously we had [empty function call](https://github.com/webiny/webiny-js/pull/3788/files/3f81da807c561eb0a215e689a1a5b842b2b5c083#diff-22f0fc472ecfa63e00eb2362b7e34e5b41b1eca5df254d624d3b9fd983188123L115), but it requires 2 arguments as defined [here](https://github.com/webiny/webiny-js/blob/next/packages/ui/src/DelayedOnChange/DelayedOnChange.ts#L7).

File: packages/app-page-builder/src/editor/Editor.tsx (hunk 9 above)
Comment (neatbyte-ivelychko) [reply packages/app-page-builder/src/editor/Editor.tsx:55]: `react-dnd v14.0.5` introduces some breaking changes that will require refactoring.

File: packages/app-page-builder/src/pageEditor/helpers.ts (hunk 23 above)
Comment (neatbyte-ivelychko) [reply packages/app-page-builder/src/pageEditor/helpers.ts:12]: Plugin name property can be `undefined` as defined [here](https://github.com/webiny/webiny-js/blob/next/packages/plugins/src/types.ts#L5). Also `plugins.byName` function accepts `undefined`, so it should not introduce any issues.

File: packages/app-page-builder/src/editor/plugins/defaultBar/components/ExportBlockButton/ExportBlockLoadingDialogContent.tsx (hunk 12 above)
Comment (neatbyte-ivelychko) [reply packages/app-page-builder/src/editor/plugins/defaultBar/components/ExportBlockButton/ExportBlockLoadingDialogContent.tsx:36]: Looks like `{message: string, code: string, data: Record<string, any>}` is query error and the error that we set in state is [PageBuilderImportExportSubTask](https://github.com/webiny/webiny-js/blob/next/packages/app-page-builder/src/types.ts#L912) error which has different properties. It’s type is also defined [here](https://github.com/webiny/webiny-js/blob/next/packages/api-page-builder-import-export/src/types.ts#L36). As we can see [here](https://github.com/webiny/webiny-js/blob/next/packages/api-page-builder-import-export/src/export/combine/formsHandler.ts#L82) it’s properties can be `{name: string, message: string, code: string}`.

File: packages/app-page-builder/src/editor/plugins/elementSettings/components/SelectField.tsx (hunk 19 above)
Comment (neatbyte-ivelychko) [reply packages/app-page-builder/src/editor/plugins/elementSettings/components/SelectField.tsx:64]: HTML’s `select` tag does not have `placeholder` property. We already use [this](https://github.com/webiny/webiny-js/blob/next/packages/app-page-builder/src/editor/plugins/elementSettings/components/SelectField.tsx#L75) approach to add a placeholder.

File: packages/app-page-builder/src/editor/plugins/elements/button/ButtonSettings.tsx (hunk 21 above)
Comment (neatbyte-ivelychko) [reply packages/app-page-builder/src/editor/plugins/elements/button/ButtonSettings.tsx]: You are correct, but it looks like [this](https://github.com/webiny/webiny-js/pull/3788/files#diff-9d98849c0deffdf9bfad3458e47ea8fdaadeb6d2f801c26504ba4a6be1d82701R44) should also be changed to string.

File: packages/app-page-builder/src/admin/views/BlockCategories/IconPicker.tsx (hunk 3 above)
Comment (brunozoric) [reply packages/app-page-builder/src/admin/views/BlockCategories/IconPicker.tsx:113]: I still dont understand why did you remove it. When I go through the code, i found that cb is sometimes sent, so it should be executed.

File: packages/app-page-builder/src/editor/Editor.tsx (hunk 9 above)
Comment (brunozoric) [reply packages/app-page-builder/src/editor/Editor.tsx:55]: Ok, let's keep the current one, if everything works.

File: packages/app-page-builder/src/editor/components/IconPicker.tsx (hunk 10 above)
Comment (brunozoric) [reply packages/app-page-builder/src/editor/components/IconPicker.tsx:189]: Same comment as few posts up.

File: packages/app-page-builder/src/admin/views/Menus/MenusDataList.tsx [hunk 24]
```diff
@@ -96,7 +96,7 @@ const PageBuilderMenusDataList = ({ canCreate }: PageBuilderMenusDataListProps)
     const slug = new URLSearchParams(location.search).get("slug");
 
     const deleteItem = useCallback(
-        item => {
+        (item: any) => {
```
Comment (brunozoric) [inline packages/app-page-builder/src/admin/views/Menus/MenusDataList.tsx]: `any`?

**Review by brunozoric — CHANGES_REQUESTED** [review]

File: packages/app-page-builder/src/admin/views/BlockCategories/IconPicker.tsx (hunk 3 above)
Comment (neatbyte-vnobis) [reply packages/app-page-builder/src/admin/views/BlockCategories/IconPicker.tsx:113]: In the original implementation, the callback function `cb` was invoked without any arguments. However, according to [this](https://github.com/webiny/webiny-js/blob/next/packages/ui/src/DelayedOnChange/DelayedOnChange.ts#L7) interface, `cb` requires two parameters: a `value` of type `string` and a callback function `cb` of type `(value: string) => void`.

Previuosly the parameters for the `onFilterChange` function were implicitly typed as `any`, which masked the issue that `cb` was being called incorrectly (without the required arguments).

I understand that `cb` might sometimes be passed and needs to be executed. However, in its previous form, it was not being used correctly as per the interface's requirement.

I see two possible solutions for this issue:
1) Refactor the `onFilterChange` function to include the `cb` parameter and use it correctly.

```
    const onFilterChange = useCallback(
        (value: string, cb?: ApplyValueCb) => {
            setFilter(value);
            if (cb) {
                cb(value, /* appropriate callback function */);
            }
        },
        [filter]
    );
```
2) Remove the `cb` function if it's confirmed to be unnecessary inside `onFilterChange`.

Comment (neatbyte-vnobis) [reply packages/app-page-builder/src/admin/views/BlockCategories/IconPicker.tsx:113]: I've just realized that these files are part of the `IconPicker`, and we're in the process of creating a new `IconPicker` component. Given that we're close to merging the new `IconPicker` and the old one will be removed, it might not be necessary to address this issue right now.

**Review by brunozoric — APPROVED** [review]
